File: server/issue-tracker-server/Models/Bug.cs
namespace issue_tracker_server.Models;

public class Bug
{
    public int BugId { get; set; }
    public required string Title { get; set; }
    public required string Description { get; set; }
    public string? ExpectedBehavior { get; set; }
    public string? Remarks { get; set; }

    public int AffectedBuildId { get; set; }
    public required Build AffectedBuild { get; set; }

    public int? FixedBuildId { get; set; }
    public Build? FixedBuild { get; set; }

    public int VariantId { get; set; }
    public required Variant Variant { get; set; }

    public int ImpactId { get; set; }
    public required Impact Impact { get; set; }

    public int StatusId { get; set; }
    public required Status Status { get; set; }

    public int ReporterId { get; set; }
    public required User Reporter { get; set; }

    public int ResponsibleId { get; set; }
    public required User Responsible { get; set; }
}

File: client/issue-tracker-ui/src/app/core/bug.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface BugFilter {
  variantIds?: number[];
  impactIds?: number[];
  statusIds?: number[];
  reporterIds?: number[];
  responsibleIds?: number[];
  affectedBuildIds?: number[];
  search?: string;
}

export interface BugResponse {
  bugId: number;
  title: string;
  description: string;
  affectedBuild: string;
  fixedBuild: string | null;
  expectedBehavior: string | null;
  remarks: string | null;
  variant: string;
  impact: string;
  status: string;
  reporter: string;
  responsible: string;
}

export interface CreateBugRequest {
  title: string;
  description: string;
  affectedBuildId: number;
  expectedBehavior: string | null;
  remarks: string | null;
  variantId: number;
  impactId: number;
  responsibleId: number;
}

export interface UpdateBugRequest {
  title: string;
  description: string;
  affectedBuildId: number;
  expectedBehavior: string | null;
  remarks: string | null;
  variantId: number;
  impactId: number;
  statusId: number;
  fixedBuildId: number | null;
  responsibleId: number;
}

@Injectable({ providedIn: 'root' })
export class Bug {
  private http = inject(HttpClient);

  getAll(filter?: BugFilter): Observable<BugResponse[]> {
    let params = new HttpParams();

    for (const [key, value] of Object.entries(filter ?? {})) {
      if (Array.isArray(value)) {
        for (const id of value) {
          params = params.append(key, id);
        }
      } else if (value) {
        params = params.set(key, value);
      }
    }

    return this.http.get<BugResponse[]>(`${environment.apiBaseUrl}/bugs`, { params });
  }

  create(request: CreateBugRequest): Observable<BugResponse> {
    return this.http.post<BugResponse>(`${environment.apiBaseUrl}/bugs`, request);
  }

  update(bugId: number, request: UpdateBugRequest): Observable<BugResponse> {
    return this.http.put<BugResponse>(`${environment.apiBaseUrl}/bugs/${bugId}`, request);
  }
}
